esp/services/ws_fs/ws_fsService.cpp

```diff
@@ -1881,49 +1881,159 @@ IPropertyTree *CFileSprayEx::getAndValidateDropZone(const char *path, const char
-                if (!startsWith(sourcePath, dropZonePlanePath))
-                    throw makeStringException(ECLWATCH_INVALID_INPUT, "Invalid source path");
-            }
-            else
-            {
-                StringBuffer s(sourcePath);
-                sourcePath.set(dropZonePlanePath);
-                addNonEmptyPathSepChar(sourcePath);
-                sourcePath.append(s);
-            }
-            getDropZoneHost(srcPlane, dropZone, sourceIPReq);
-            sourcePlaneReq.append(srcPlane);
+    if (!isPathSepChar(fileReq[0]) | (fileReq[0] != fileReq[1]))
+        return;
+
+    splitUNCFilename(fileReq, &hostInPathStr, &localPath, &localPath, &localPath);
+    const char* hostInPath = hostInPathStr.str() + 2;
+    if (sourceHostEp)
+    {
+        SocketEndpoint hostInPathEp(hostInPath);
+        if (!sourceHostEp->ipequals(hostInPathEp))
+        {
+            VStringBuffer msg("The host %s defined in the SourcePath does not match with the host %s defined in SourceIP. ", hostInPath, sourceHostReq);
+            msg.append("A dropzone name specified in the SourcePlane is preferred. The host/IP should not be contained in the SourcePath.");
+            throw makeStringException(ECLWATCH_INVALID_INPUT, msg.str());
+        }
+    }
+    else if (!isHostInPlane(dropZone, hostInPath, true))
+        throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Host %s in %s is not defined within the dropzone %s.", hostInPath, fileReq, dropZoneName);
+}
+
+static void ensureDZFileScopePermission(IEspContext& context, const char* plane, const char* host, const char* path, SecAccessFlags permissionReq)
+{
+    SecAccessFlags permission = getDZFileScopePermissions(context, plane, path, host);
+    if (permission < permissionReq)
+        throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Access DropZone Scope %s %s not allowed for user %s (permission:%s). Read Access Required.",
+            plane, path, context.queryUserId(), getSecAccessFlagName(permission));
+}
+
+void CFileSprayEx::readAndCheckSpraySourceReq(IEspContext& context, const char* srcHost, const char* srcPath, const char* srcPlane,
+    StringBuffer& sourcePlaneReq, StringBuffer& sourceHostReq, StringBuffer& sourcePathReq)
+{
+    //Clean srcPlane and srcHost
+    sourcePlaneReq.set(srcPlane);
+    sourceHostReq.set(srcHost);
+    sourcePlaneReq.trim();
+    sourceHostReq.trim();
+    if (sourcePlaneReq.isEmpty() && sourceHostReq.isEmpty())
+        throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Neither source plane nor source host supplied. A dropzone name specified in the SourcePlane is preferred.");
+
+    //Clean srcPath
+    StringBuffer sourcePath, s(srcPath);
+    s.trim();
+    if (s.isEmpty())
+        throw makeStringException(ECLWATCH_INVALID_INPUT, "Source path for spraying a file not specified.");
+    if (!sourcePlaneReq.isEmpty())
+        getStandardPosixPath(sourcePath, s.str());
+    else
+        getStandardPosixPath(sourcePathReq, s.str());
```

Comment:
sorry, I don't understand the logic here..
If no plane in request, then fill incoming srcPath, into [outbound] sourcePathReq, leaving 'sourcePath' empty,
which is then used below (line 1935)..

Needs comments to clarify at least.